lti_consumer/plugin/views.py

```diff
@@ -231,3 +233,68 @@ def scores(self, request, *args, **kwargs):
             status=status.HTTP_201_CREATED,
             headers=headers
         )
+
+
+class LtiNrpsViewSet(viewsets.ViewSet):
+    """
+    Api Endpoints for LTI NPRS Services
+    """
+    serializer_class = LtiNrpsContextMembershipSerializer
+    pagination_class = None
+
+    # Custom permission classes for LTI APIs
+    authentication_classes = [Lti1p3ApiAuthentication]
+    permission_classes = [LtiNrpsPermissions]
+
+    # Renderer classes to accept LTI NRPS content types
+    renderer_classes = [
+        MembershipResultRenderer,
+    ]
+
+    @action(methods=['GET'], detail=False, url_path='memberships')
+    def memberships(self, request, *args, **kwargs):
+        """
+        LTI NRPS Context Membership service endpoint.
+
+        See full API spec on -
+        https://www.imsglobal.org/spec/lti-nrps/v2p0#context-membership
+        """
+        course_key = request.lti_configuration.location.course_key
+
+        # get all course enrollments
+        enrollments = compat.get_user_enrollments(course_key)
+        enrolled_users = [enrol.user for enrol in enrollments]
+
+        # get user profiles in bulk
+        user_profiles = compat.get_user_profiles(enrolled_users)
+
+        # get external ids in bulk
+        user_eids = compat.get_external_ids(enrolled_users)
```

Comment:
Instead of fetching everything all at once, I would only fetch what's requested by the API call.

Does the spec has any filtering per user_id/role/limit? If so, we should implement it, otherwise each NRPS API call is going to be costly.

Reply:
@giovannicimolin, Yes, there is three different thing needs to be implemented here. 
- Filtering via [role](https://www.imsglobal.org/spec/lti-nrps/v2p0#role-query-parameter)
- Pagination via [Next URL](https://www.imsglobal.org/spec/lti-nrps/v2p0#limit-query-parameter). It seems how we do pagination is up to us, we can use ID or some other param here.
- Filtering via [difference](https://www.imsglobal.org/spec/lti-nrps/v2p0#membership-differences)? 

My current implementation accommodates only Students. But the Spec also speaks about other roles. ``CourseEnrollment`` doesn't seem to have Teacher / Staff user linked to it. Is there any better Model? Otherwise, this pagination implementation will be tricky.

Reply:
@giovannicimolin, The API now supports filtering and pagination.